routers/api/v1/repo/pull.go

```diff
@@ -759,13 +759,70 @@ func MergePullRequest(ctx *context.APIContext, form auth.MergePullRequestForm) {
 		return
 	}
 
-	if !pr.CanAutoMerge() {
-		ctx.Error(http.StatusMethodNotAllowed, "PR not in mergeable state", "Please try again later")
+	if pr.HasMerged {
+		ctx.Error(http.StatusMethodNotAllowed, "PR already merged", "")
 		return
 	}
 
-	if pr.HasMerged {
-		ctx.Error(http.StatusMethodNotAllowed, "PR already merged", "")
+	// handle manually-merged mark
+	if models.MergeStyle(form.Do) == models.MergeStyleManuallyMerged {
+		prUnit, err := pr.BaseRepo.GetUnit(models.UnitTypePullRequests)
+		if err != nil {
+			ctx.ServerError("pr.BaseRepo.GetUnit(models.UnitTypePullRequests): %v", err)
+			return
+		}
+		prConfig := prUnit.PullRequestsConfig()
+
+		// Check if merge style is correct and allowed
+		if !prConfig.IsMergeStyleAllowed(models.MergeStyleManuallyMerged) {
+			ctx.Error(http.StatusMethodNotAllowed, "manually-merged is not allowed", "")
+			return
+		}
+
+		if len(form.MergeCommitID) < 40 {
+			ctx.Error(http.StatusMethodNotAllowed, "commit id is wrong, need real commit id on base branch", "")
+			return
+		}
+
+		commit, err := ctx.Repo.GitRepo.GetCommit(form.MergeCommitID)
+		if err != nil {
+			ctx.ServerError("ctx.Repo.GitRepo.GetCommit", err)
+			return
+		}
+
+		branchName, err := commit.GetBranchName()
+		if err != nil {
+			ctx.ServerError("commit.GetBranchName", err)
+			return
+		}
+
+		if branchName != pr.BaseBranch {
+			ctx.Error(http.StatusMethodNotAllowed, "commit id is wrong, need real commit id on base branch", "")
+			return
+		}
```

Comment:
So this is unnecessarily strict I think.

You could reasonably argue that any commit on the base branch is a reasonable place to say this was merged to.

There are several ways to detect that:

```
git branch --contains COMMIT_ID BRANCH_NAME
```

where if it returns anything then the commit is on the branch

or:

```
git rev-list BRANCH_NAME | grep COMMIT_ID
```

would also work.